commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+				return nil, fmt.Errorf("found different roots (%x, %x) for the same source chain %d",
+					existingRoot, lu.Root, lu.LaneSource.SourceChainSelector)
+			}
+		}
+	}
+
+	return &ReportSignatures{
+		Signatures:  ecdsaSignatures,
+		LaneUpdates: fixedDestLaneUpdates,
+	}, nil
+}
+
+func (c *PBClient) listenForRmnReportSignatures(
+	ctx context.Context,
+	requestIDs mapset.Set[uint64],
+	reportSigReq *rmnpb.ReportSignatureRequest,
+	signersRequested mapset.Set[uint32],
+) ([]*rmnpb.EcdsaSignature, error) {
+	tReportsInitialRequest := time.NewTimer(c.reportsInitialRequestTimerDuration)
+	reportSigs := make([]*rmnpb.ReportSignature, 0)
+	finishedRequests := mapset.NewSet[uint64]()
+	respChan := c.rawRmnClient.Recv()
+
+	for {
+		select {
+		case resp := <-respChan:
+			responseTyp, err := c.parseResponse(&resp, requestIDs, finishedRequests)
+			if err != nil {
+				c.lggr.Errorw("failed to parse RMN response", "err", err)
+				continue
+			}
+
+			reportSig := responseTyp.GetReportSignature()
+			if reportSig == nil {
+				c.lggr.Errorf("RMN node returned an unexpected type of response: %+v", responseTyp.Response)
+			} else {
+				reportSigs = append(reportSigs, reportSig)
+			}
+
+			if len(reportSigs) >= c.minSigners {
+				ecdsaSigs := make([]*rmnpb.EcdsaSignature, 0)
+				for _, rs := range reportSigs {
+					ecdsaSigs = append(ecdsaSigs, rs.Signature)
+				}
+				return ecdsaSigs, nil
+			}
+		case <-tReportsInitialRequest.C:
+			// Send to the rest of the signers
+			for node := range randomIter(c.rmnNodes) {
+				if !node.IsSigner || signersRequested.Contains(node.ID) {
+					continue
+				}
+				req := &rmnpb.Request{
+					RequestId: newRequestID(),
+					Request: &rmnpb.Request_ReportSignatureRequest{
+						ReportSignatureRequest: reportSigReq,
+					},
+				}
+				if err := c.marshalAndSend(req, node.ID); err != nil {
+					c.lggr.Errorw("failed to send report signature request", "node_id", node.ID, "err", err)
+					continue
+				}
+				requestIDs.Add(req.RequestId)
+				signersRequested.Add(node.ID)
+			}
+		case <-ctx.Done():
+			return nil, ErrTimeout
+		}
+	}
+}
+
+func (c *PBClient) ensureEnoughSignedObservations(rmnSignedObservations []rmnSignedObservationWithMeta) error {
+	counts := make(map[uint64]int)
+	for _, so := range rmnSignedObservations {
+		for _, lu := range so.SignedObservation.Observation.FixedDestLaneUpdates {
+			counts[lu.LaneSource.SourceChainSelector]++
+		}
+	}
+
+	for chain, count := range counts {
+		if count < c.minObservers {
+			return fmt.Errorf("not enough observations for chain=%d count=%d minObservers=%d",
+				chain, count, c.minObservers)
+		}
+	}
+
+	return nil
+}
+
+type RawRmnClient interface {
+	Send(rmnNodeID uint32, request []byte) error
+	// Recv returns a channel which can be used to listen on for
+	// responses by all RMN nodes. This is expected to be monitored
+	// by the plugin in order to get RMN responses.
+	Recv() <-chan RawRmnResponse
+}
```

Comment:
Does send and Recv need to be separated? I think the code would be simpler if you pass the response channel into the send function. Alternatively you could make the channel in the send function and return it for each send, but then you'd have to read each channel.

Reply:
This interface was given to me as is, but I believe we can make such changes.
@makramkd @gtklocker ?

Reply:
The raw client interface is not set in stone, but I think we need to actually implement this client to see what is possible. We can come back and change it if we see that passing in a chan is doable and easy on the implementation side

Reply:
I believe that the thinking behind `Recv` is you can call it "anytime" (i.e before you even send any requests). It can be considered the channel that ALL RMN comms arrive at and will have the same lifetime as the RMN raw client itself.

Reply:
The more I think about it, I the less sense it makes to have concurrency in the client at all. Follow the pattern used in the go http client, this code can focus on communication. Since communication is localized to a single function you don't need anything more complicated than a go routine to send the requests and put the response in a go routine, and an inline range to process results.

```
respChan := make(chan RawRMNResponse)
go func() {
    for id, req := range requests {
        resp, err <- Send(ctx, id, req)  // the ctx is also missing here.
        if err != nil {
            cloe(respChan) // or the err group thingie
        }
    }
}

var results []RawRMNResponse
for result := range respChan {
    results = append(results, result)
}

aggregate(results)
```

Reply:
Atm, this interface did fit nice in the approach and the testing. We can revisit in a follow-up